build_runner/lib/src/asset_graph/graph.dart

```diff
@@ -191,7 +191,10 @@ class AssetGraph {
       var builderOptionsNode = get(node.builderOptionsId);
       builderOptionsNode.outputs.remove(id);
     }
-    _nodesByPackage[id.package].remove(id.path);
+    // Synthetic nodes need to be kept to retain dependency tracking.
+    if (node is! SyntheticSourceAssetNode) {
+      _nodesByPackage[id.package].remove(id.path);
```

Comment:
Note that we are still adding thise node to `removedIds`... is that correct?

Reply:
Good question...

We apparently also explicitly intended for SyntheticSourceAssetNodes to be removed:

https://github.com/dart-lang/build/blob/b7bea8daa7f8fecc866e2b98b20edeb513a6c548/build_runner/lib/src/asset_graph/graph.dart#L449-L451

We count on these removed nodes being tracked so we can remove them from the set of assets consider inputs through a series of phases

https://github.com/dart-lang/build/blob/b7bea8daa7f8fecc866e2b98b20edeb513a6c548/build_runner/lib/src/asset_graph/graph.dart#L419

`allInputs` there starts as the set of sources in the environment, not in the graph, so the synthetic nodes wouldn't be in that set. I _think_ this is safe as is.